feat: Verify Client Certificate

sklirg opened the pull request:
Depends on #142.

Fixes #81 

Only includes verification checks of DNS names.

sklirg committed:
feat: Verify Client Certificate based on the presence of an expected SAN DNS Name

sklirg commented:
One interesting caveat: This verification has to happen before "HTTP", which means that we don't know what URL the user is accessing at the point of verification. In turn, this means that our current "auth"-dict which is based on `/path`s will not work for verifying using client certificates.

Potentially, we'll have to create separate http receivers for each handler. It's not a problem, really, it just doesn't work as well with our existing auth-scheme.

sklirg commented on receiver/http.go:
```diff
@@ -56,6 +57,7 @@ type HTTP struct {
 	Certfile      string                        `doc:"Path to certificate file for TLS. If left blank, un-encrypted HTTP is used."`
 	Keyfile       string                        `doc:"Path to key file for TLS."`
 	AcceptableCAs []string                      `doc:"Paths to files containing CAs which are accepted for Client Certificate authentication."`
+	SANDNSName    string                        `doc:"DNS name which has to be present in SAN extension of x509 certificate when using Client Certificate authentication"`
```
We could try to work this more into the style of HTTPAuth (line 42-46), as adding struct fields for each of the potential ways to verify (including DNSNames, EmailAddresses, IPAddressess, URIs) could become a bit messy if all added onto the base HTTP struct.

Just a thought.

sklirg commented on receiver/http.go:
```diff
@@ -192,6 +198,43 @@ func (htt *HTTP) Start() error {
+// during TLS handshake by comparing its extensions (such as SAN) to
+// some expected value(s)
+func (htt *HTTP) verifyPeerCertificate(_ [][]byte, verifiedChains [][]*x509.Certificate) error {
+	certLogger := httpLog
```
Logs in here should maybe be left out of the common path, i.e. opt-in through config? This could be a high throughput function which shouldn't incur any extra overhead unless requested.

KristianLyng commented on receiver/http.go:
```diff
@@ -192,6 +198,43 @@ func (htt *HTTP) Start() error {
+			continue
+		}
+		cert := chain[0]
+		certLogger := certLogger.WithFields(log.Fields{
```
I am not a big fan of overshadowing variable names O_o. The logger isn't a big issue, but imagine trying to debug this at 3AM and not understanding why certLogger means something different on line 206 and line 222.

KristianLyng reviewed (changes requested):
Right, so the big thing about this is obviously the lack of granularity.

AFAICT, there's not cryptographic reason why we can't check the SAN-DNS connection after we've got the request. We should then have validated the certificate and established a TLS session both ways. 

However, I realize that this implementation is by far the most trivial since it basically offloads the entire job to the standard library.

However, did you investigate if we can authorize in ServeHTTP instead? I see that we have http.Request there, which in turn has tls.ConnectionState which IN turn has the same verifiedChains array that verifyPeerCertificate relies on. Perhaps we can simply call it manually in ServeHTTP with relevant path(s) instead of relying on it being called implicitly? At first glance, that doesn't seem like too much effort, and would allow us to have path-specific authorization again.

As for how to configure that, I'm thinking just adding SANDNSName to HTTPAuth is an alternative, though it requires some more Validate() work to be fool proof.

sklirg committed:
feat: Engage VerifyPeerCertificate only on configured paths

Also, if a Client Certificate is presented upon connecting, we verify
the certificate to be valid for one of the paths published.

sklirg commented:
I moved the verification logic into the auth handler, as is done for basic auth. It seems to play nice.

KristianLyng reviewed (approved):
This looks good to me now.
